source

```
@@ -80771,6 +80987,918 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
+   data-x="dom-hidePopover">hidePopover</span>()</code></dt>
+
+   <dd>Hides the popover <var>element</var> by removing it from the top layer and applying
+   <code data-x="">display: none</code> to it.</dd>
+
+   <dt><code data-x=""><var>element</var>.<span subdfn
+   data-x="dom-togglePopover">togglePopover</span>()</code></dt>
+
+   <dd>If the popover <var>element</var> is not showing, then this method shows it. Otherwise, this
+   method hides it.</dd>
+  </dl>
+
+  <p>The <dfn method for="HTMLElement"><code data-x="dom-showPopover">showPopover()</code></dfn>
+  method steps are:</p>
+
+  <ol>
+   <li><p>If the result of running <span>check popover validity</span> given <span>this</span> and
+   false is false, then throw an <span>"<code>InvalidStateError</code>"</span>
+   <code>DOMException</code>.</p></li>
+
+   <li><p>Run <span>show popover</span> given <span>this</span>.</p></li>
+  </ol>
+
+  <p>To <dfn>show popover</dfn>, given an <span data-x="HTML elements">HTML element</span>
+  <var>element</var>:</p>
+
+  <ol>
+   <li><p>If <var>element</var> is a <code>dialog</code> element and has an <code
+   data-x="attr-dialog-open">open</code> attribute, then throw an
+   <span>"<code>InvalidStateError</code>"</span> <code>DOMException</code>.</p></li>
+
+   <li><p>If <var>element</var> has the <span>fullscreen flag</span> set, then throw an
+   <span>"<code>InvalidStateError</code>"</span> <code>DOMException</code>.</p></li>
+
+   <li><p><span>Assert</span>: <var>element</var> is not in <var>element</var>'s <span>node
+   document</span>'s <span>top layer</span>.</p></li>
+
+   <li><p>If the result of <span data-x="concept-event-fire">firing an event</span> named <code
+   data-x="event-popovershow">popovershow</code> at <var>element</var> is false, then
+   return.</p></li>
+
+   <li>
+    <p>If the result of running <span>check popover validity</span> given <var>element</var> and
+    false is false, then return.</p>
```

Already made this suggestion, but accidentally deleted it:

```suggestion
    false is false, then throw an <span>"<code>InvalidStateError</code>"</span>
    <code>DOMException</code>.</p>
```

---

Since we already checked and threw exceptions earlier when showPopover is called, I don't think this one should throw an exception. The implementation in chromium doesn't.

---

We can only reach this point of the first call to "check popover validity" succeeded and no exception was thrown. For both `showPopover()` and `hidePopover()` the current PR (squash and rebase 34) first does the check and throws exceptions if it fails, and does it again after events are fired, but if it fails just silently returns. There are 4 such call sites that just to "then return" but I only suggested it here.

For `hidePopover()` I think there's a good argument for silent return: if the event handler just removed the element, it's still hidden. The focus restoring logic won't run, but if that's a problem we can change the algorithm and make that work, while still not throwing an exception.

For `showPopover()` I don't understand the reason though. If the event handler does something that now makes it impossible to show the popover, the operation can't be finished, so isn't an exception appropriate?

I think this isn't cut and dry, it's a judgment call, but I don't quite get the reason for it.

---

@mfreed7 is it feasible to throw an exception here?

---

To be more specific, this is where we are talking about throwing an exception instead of returning: https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/html_element.cc;l=1328-1331;drc=8ffd224064264fafa77ab24cfcc0fb1584094f30

---

As with the equivalent cases in `hidePopover()`, there are places that get called where an exception can't be thrown. For example, from [here](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/forms/html_form_control_element.cc;l=409;drc=a98d92c84d9821c8de97618095ed9df220cd1da7) (the case when an invoking element like a button triggers the popover). The general problem with all of these cases, which happen *after* an event handler runs, is that all of the state can be changed by the event handler. Before the event handler, it makes sense to throw exceptions, since the pre-conditions can be checked. After event handlers, that's not the case.

---

In that case, it sounds like this algorithm is not reusable, and you need a different algorithm for `hidePopover()` (which needs to throw exceptions) vs. general popover-hiding (which cannot).

---

...so should we have a "throwExceptions" flag passed to everything that can throw exceptions which is set to true from the DOM method and false from everything else? @mfreed7 would that work? I'm not sure which event handlers you're talking about - do you mean that we can never throw an exception after firing an event?

---

If I understand this suggestion, all that would be required is to split the existing hidePopover algorithm into two pieces, one before the event is fired and one after. And the former calls the latter. And then elsewhere (e.g. in the light dismiss routines) you only call the second one, which doesn't throw.

E.g.
```
hide popover 1:
 1. If error conditions, throw exception
 2. Fire hide event.
 3. Call hide popover 2.

hide popover 2:
 1. If error conditions, return.
 2. Do the rest...

light dismiss algorithm:
 1. If we want to light dismiss popover, call hide popover 2 with popover.
```

---

I rewrote this algorithm based on mason's chromium patch